zrml/global-disputes/src/migrations.rs

```diff
@@ -0,0 +1,398 @@
+                        .expect("Owners len is 1, but could not get this owner.")
+                        .clone(),
+                    fee: T::VotingOutcomeFee::get(),
+                }),
+                _ => Some(Possession::Shared { owners }),
+            };
+
+            let outcome_info =
+                OutcomeInfo { outcome_sum: winner_info.outcome_info.outcome_sum, possession };
+            let gd_info = GDInfo {
+                winner_outcome: winner_info.outcome,
+                outcome_info,
+                status: GDStatus::Finished,
+            };
+            crate::GlobalDisputesInfo::<T>::insert(market_id, gd_info);
+            total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+        }
+
+        // Look here https://github.com/paritytech/substrate/blob/7ba1ef526fbcc224d3bd446c83ee66a83cacec49/frame/assets/src/migration.rs#L69-L79
+        let mut translated = 0u64;
+        Outcomes::<T>::translate::<OldOutcomeInfo<BalanceOf<T>, OwnerInfoOf<T>>, _>(
+            |_key1, _key2, old_value| {
+                translated.saturating_inc();
+
+                let owners = old_value.owners;
+                let owners_len = owners.len();
+                let possession = match owners_len {
+                    0usize => None,
+                    1usize => Some(Possession::Paid {
+                        owner: owners
+                            .get(0)
+                            .expect("Owners len is 1, but could not get this owner.")
+                            .clone(),
+                        fee: T::VotingOutcomeFee::get(),
+                    }),
+                    _ => Some(Possession::Shared { owners }),
+                };
+
+                let new_value = OutcomeInfo { outcome_sum: old_value.outcome_sum, possession };
+
+                Some(new_value)
+            },
+        );
+        log::info!("ModifyGlobalDisputesStructures: Upgraded {} outcomes.", translated);
+        total_weight = total_weight
+            .saturating_add(T::DbWeight::get().reads_writes(translated + 1, translated + 1));
+
+        StorageVersion::new(GD_NEXT_STORAGE_VERSION).put::<GDPallet<T>>();
+        total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+        log::info!("ModifyGlobalDisputesStructures: Done!");
+        total_weight
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<(), &'static str> {
+        use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+        let mut counter = 0_u32;
+        for (market_id, winner_info) in crate::Winners::<T>::iter() {
+            Self::set_temp_storage(winner_info, &format!("{:?}", market_id));
+
+            counter = counter.saturating_add(1_u32);
+        }
+        let counter_key = "counter_key".to_string();
+        Self::set_temp_storage(counter, &counter_key);
+        Ok(())
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade() -> Result<(), &'static str> {
```

Comment:
What would you suggest additionally? What do you mean with "the outcome token is still the correct type"? It feels unclear to me what to remove (do we even should remove it..?) and what to add.

Reply:
> What would you suggest additionally?

I think you ask this everytime I criticize the `post_upgrade` implementation and I never have a good answer. ;)

The problem with the `post_upgrade` is that it basically forces you to programatically handle the test data. Basically, if you want to test a function `f(value: u32) -> u32` which returns two times `value` plus 1, then you take a couple of examples for `value` and the expected value, e.g. `(2, 5)` or `(0, 1)` and check if the expected value is the actual value.

But `post_upgrade` basically hands you three numbers `(value, actual)` (where `value` is the old market and `actual` is the new market; and `f` is the migration, of course) and wants you to verify that `f(value)` is correct. But there's no good way to do this, because we don't know what the correct expected value is without calculating it from `value` first. But if we use `f` to calculate what the expected value is, then all we do is verify that `f` is deterministic, i.e. running `f` with the same input twice results in the same output. That's not ideal (but it doesn't hurt us either).

One solution is to implement another version of `f`, let's call it `g` and then test that both implementations yield the same result. If two vastly different implementations yield the same result, it's quite likely that they're both right. A great example for this is how many linear algebra libraries test against data that other linear algebra libraries have produced. I'm pretty certain that this is a bad fit for testing a migration.

The other is to make sanity tests. Does the result `f(value)` even remotely make sense? For example, if you know that `value` is always less than `1_000`, then you can test that `f(value) < 2_001`. That's what I'm suggesting here. What can we do in this particular case...?

- Ensure that if a fee was paid, then it is equal to `VotingOutcomeFee`.
- If votes were made, then the locked amount is correct.
- If there's a winner, then it actually has the highest number of votes.

Reply:
I understand your approach here. It seems a little bit overkill for empty storage items I must admit. Are you okay with leaving it how it is and I mark your words here in my mind for future migrations?

Reply:
Yeah, that's fine.